arch/x86/core/intel64/irq.c

```diff
@@ -156,3 +156,59 @@ void arch_sched_ipi(void)
 	z_loapic_ipi(0, LOAPIC_ICR_IPI_OTHERS, CONFIG_SCHED_IPI_VECTOR);
 }
 #endif
+
+/* The first bit is used to indicate whether the list of reserved interrupts
+ * have been initialized based on content stored in the irq_alloc linker
+ * section in ROM.
+ */
+#define IRQ_LIST_INITIALIZED 0
+
+static ATOMIC_DEFINE(irq_reserved, CONFIG_MAX_IRQ_LINES);
+
+static void irq_init(void)
+{
+	extern uint8_t __irq_alloc_start[];
+	extern uint8_t __irq_alloc_end[];
+	const uint8_t *irq;
+
+	for (irq = __irq_alloc_start; irq < __irq_alloc_end; irq++) {
+		__ASSERT_NO_MSG(*irq < CONFIG_MAX_IRQ_LINES);
+		atomic_set_bit(irq_reserved, *irq);
+	}
+}
+
+unsigned int arch_irq_allocate(void)
```

Comment:
If I am reading this correctly, there is a potential race condition in this routine.

At the present time, I do not observe anything preventing two threads from calling arch_irq_allocate() at the same time, or stating that it should not be done. Imagine two threads--A and B. A is the first to call arch_irq_allocate() and in turn calls irq_init(). However, if thread B gets scheduled and interrupts the initialization of irq_reserved, it can incorrectly allocate an IRQ that was statically reserved but not yet officially marked as such.

Can such a scenario occur?

Reply:
I could document that it requires the caller to disable the irq. 

I assumed all drivers did that at initialization, but it seems that it's not always the case: ns16550 does not for instance... That's a bug.